### PR TITLE
feat(#715): root implicit AD for C4v CTMRG (Phase 0)

### PR DESCRIPTION
**Pull request by yingjerkao**
> 🤖 **AI-generated PR** — written by Claude Code, posted by @yingjerkao.

Phase 0 of #715. Implements §III of [arXiv:2607.15030](https://arxiv.org/abs/2607.15030) (Burgelman, Francuz, Brehmer, Devos, Haegeman, Verstraete, Vanhecke) for the dense C4v path.

## What changes

New module `src/tenax/algorithms/_ctm_c4v_root_implicit.py`. Nothing existing is touched — the forward reuses `_c4v_sweep` unmodified, so there is no risk to the production path.

The production backward characterises the converged environment as a **fixed point** of one CTM sweep and solves `(1 - J^T) λ = dE/denv`, differentiating the sweep and its truncated-decomposition backward. This module characterises it as the **root** of an algebraic equation `F(y, p) = 0` over `y = (C, E, u)` and solves `F̆ ∂_y F = ў` (paper Eqs. 14–18):

```
R_C = U† M U - λ_C C                  (Eq. 26)
R_E = U† N U - λ_E E                  (Eq. 27)
R_u = (U_perp† M U) C*^-1 - λ_C u     (Eq. 28)
```

with `U = U* + U_perp u` (Eq. 31) — the isometry varies only along its null space, so the in-subspace rotations that produce the `1/(λ_i - λ_j)` divergence never enter. The `u` block is conditioned by the *truncation gap* instead of by gaps inside the kept block.

`F` is a handful of contractions, so **no `eigh` or `svd` backward appears in the gradient path at all**. `test_no_eigh_or_svd_in_the_differentiated_equations` asserts that structurally on the jaxpr rather than taking it on faith.

## Two things the paper leaves implicit

Both cost a debugging cycle, so they are documented in the module:

1. **`(C, E, U)` must be mutually self-consistent**, not merely converged in some other basis. Eq. 28 is amplified by `C*^-1`, which turned a 1e-12 basis mismatch into a residual of 3.3e3 in the first draft. `c4v_root_parametrize` now iterates the (enlarged corner → `eigh` → project) map to self-consistency.
2. **That loop oscillates unless the isometry gauge is pinned** — `eigh` flips column signs and freely rotates near-degenerate eigenvectors, so the residual alternated between 1e-13 and 1.5e-2. Procrustes alignment against the previous step fixes it. Forward-pass bookkeeping only; the backward never sees it.

## Gate

Directional finite-difference parity, `D=2`, dense, `eigh` projectors:

| case | ‖F(y*)‖ | adjoint resid | best rel. FD error |
|---|---|---|---|
| XXZ Δ=0.3, χ=6 | 9.6e-11 | 2.5e-13 | 8.5e-10 |
| Heisenberg Δ=1.0, χ=6 | 6.3e-13 | 3.2e-11 | 1.6e-09 |
| Heisenberg Δ=1.0, χ=4 | 5.5e-12 | 1.2e-12 | 4.9e-08 |
| XXZ Δ=1.0, χ=8 | 4.7e-12 | 1.2e-10 | 1.7e-08 |

Δ=1.0 is the isotropic point where the corner spectrum is SU(2)-degenerate. No Lorentzian regularisation is used anywhere in the backward.

`root_residual` is reported and warned on: per the paper's Fig. 1 it is what bounds gradient accuracy, so it must not be silently absorbed. Truncating through a numerically degenerate spectrum (near-product state at χ=8) raises the residual to ~6e-2 — the guard fires, and the gradient stays finite rather than NaN-ing.

## Scope

Dense + C4v only. `SymmetricTensor` raises `TypeError` pointing at Phase 3. Phases 1–3 (asymmetric CTMRG, unit cells, block-sparse) are tracked in #715.

8 new tests, `algorithm` bucket, 43 s.

### CODE REVIEW

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Both findings were valid and are fixed in 2337db6.

**Reject unconverged adjoint solves.** Agreed, and the framing in the review is
the right one: an unconverged solve does not give an approximate gradient, it
gives an arbitrary one, because `F̆` is defined as *the* solution of Eq. 17.
The residual was computed but surfaced only under `return_diagnostics`, so the
default path returned a finite, plausible-looking, wrong gradient in silence.
Now checked unconditionally against a new `solve_residual_warn`.

Writing the test for this turned up a second, smaller problem: JAX's GMRES
counts `maxiter` in **restarts**, not iterations, so `solve_maxiter=1` still
runs a full 30-dimensional Krylov space and converges fine. The original
warning text ("after N iterations") was misleading about which knob to turn.
It now names both `solve_maxiter` and `solve_restart`, and the test squeezes
both.

**Skip the adjoint solve in the energy-only helper.** Agreed — it delegated to
`_and_grad` and discarded the gradient, so every energy-only caller paid for a
VJP plus a solve, and could fail on a solve it never asked for. It now runs
forward → parametrise → energy only. Pinned by a test that sabotages
`_solve_root_adjoint` and asserts the energy still comes back, matching the
`_and_grad` value to 1e-12.

File: src/tenax/algorithms/_ctm_c4v_root_implicit.py [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+    _, vjp_p = jax.vjp(F_of_p, A_data)
+    grad_indirect = vjp_p(F_bar)[0]
+
+    grad = grad_direct - grad_indirect
+
+    if return_diagnostics:
+        diag = {
+            **meta,
+            "root_residual": root_residual,
+            "adjoint_residual": float(solve_resid),
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/_ctm_c4v_root_implicit.py:586]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Verified **fixed** on `main`. `_ctm_c4v_root_implicit.py` now checks the adjoint residual unconditionally rather than only under `return_diagnostics`, and the in-code comment records the exact scenario from this report ("``solve_maxiter=1`` returns a finite, badly wrong gradient in silence"). The warning names the residual, the tolerance, the restart count and the remedy.

One carry-over: that check is written `if float(solve_resid) > solve_residual_warn:`, which is the NaN-fail-open shape — `nan > x` is `False`, so a NaN adjoint residual takes the silent branch. Folded into **#796**, which covers the same comparison shape elsewhere in this file. Resolving this thread on the substance.

[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/_ctm_c4v_root_implicit.py:586]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Marked outdated by GitHub and not re-verified against current `main` — the energy-only helper has been restructured since. Recording the concern rather than silently dropping it: if a forward-only caller (line search, diagnostics) still constructs both VJPs and runs the GMRES adjoint before discarding the gradient, that is real cost on the hottest path, and it can fail on an adjoint the caller never asked for. Worth a look next time this file is opened; not carried into an issue because I could not confirm it still applies.
